### ISSUE TITLE
Application crashes when accessing fields in API.swift

### ISSUE THREAD

**Issue by bmunge** [apollographql/apollo-ios#530]
I'm having issues with the double optionals on many of the generated fields. When an object is instantiated and I attempt to access a value I get the following crash:

```
Could not cast value of type 'Swift.Optional<Swift.String>' (0x11814aa90) to 'Swift.String' (0x11812f888).
```

The code in question is the get block here:

```
  /// Filter results by user provided string
  public var propertyName: Swift.Optional<String?> {
    get {
      return graphQLMap["propertyName"] as! Swift.Optional<String?>
    }
    set {
      graphQLMap.updateValue(newValue, forKey: "propertyName")
    }
  }
```

My `graphQLMap` looks fine with an optionally wrapped nil value in place for this value. Here's the output for this key:

```
  ▿ 10 : 2 elements
    - key : "propertyName"
    ▿ value : Optional<JSONEncodable>
      - some : nil
```

These double optionals are creating a lot of issues for me and make any code using them kind of ugly with the additional unwrapping logic. Is there a way to exclude this double optional approach to generation (along with force casts in the get's) for output?

**Comment by designatednerd** [comment]
@martijnwalraven Do you have some background as to why this is using double optionals here?

**Comment by bmunge** [comment]
@martijnwalraven given insight in other issues:

> The reason for the double optionals is to be able to differentiate between a value not being present and a value being set to `null`, which is needed because GraphQL [considers these two cases to be different](http://facebook.github.io/graphql/October2016/#sec-Null-Value). In a mutation that updates an object for example, not providing a value for a field could mean not modifying that field, while `null` would clear the current value.
> 
> The nice thing about these double optionals is that you don't need to be aware of them unless you're providing an explicit `null`:
> 
> ```swift
> input.caption = "foo"
> input.caption = nil // don't include field
> input.caption = .some(nil) // include explicit null
> ```

I understand that approach and agree, however I think if possible we should address the crash that occurs when accessing these parameter objects in the client.

**Comment by designatednerd** [comment]
Ahhhh got it (sorry, second day on the project 😛).  To be clear, this is crashing in the getter, on the `as!`, correct? Do you have any context in terms of what you're trying to do with the value - hand it to UI or some other structure, or something else?

**Comment by bmunge** [comment]
Yep, the crash is occurring on the get block.

We have a page building various inputs that ultimately populate a single object that is used as a query parameter. I was attempting to use the parameter object to be able to set and change values as the user interacted with the UI before sending the query. Additionally, we often need to get the value to perform some client side validation.

**Comment by designatednerd** [comment]
@martijnwalraven Correct me if I'm wrong but from our discussion earlier today this doesn't sound like the use case this was designed for - it sounded like this was more designed to be an `as!` since it's using things which are validated at compile time as valid for a request rather than being used for a builder-type object. 

@bmunge Can you discuss a little why you'd like to use this object to build up the query param rather than building it up in a separate object? It'd probably help us quite a bit with debugging.

**Comment by bmunge** [comment]
That's the solution I ultimately went with, but I found it odd that using the getter always resulted in a crash. It seems like the underlying object being retrieved supported the structure so I didn't expect to see a crash.

**Comment by designatednerd** [comment]
Yeah, I think the architecture as it stands is more for create-then-send rather than ongoing usage. It still probably shouldn't crash though 🙃

**Comment by alistra** [comment]
`  public var completedAt: Swift.Optional<DateTime?> {
    get {
      return graphQLMap["completedAt"] as! Swift.Optional<DateTime?>
    }
    set {
      graphQLMap.updateValue(newValue, forKey: "completedAt")
    }
  }
`

We're having the same issue with a double optional property. We're getting crashes when accessing the property.

We use the getter to serialize the input and send it later in case of a network failure.

**Comment by designatednerd** [comment]
Definitely related to #606 - I'll try to get a look at this next week. At the very least we shouldn't be force-unwrapping there.

**Comment by designatednerd** [comment]
Hello! [Tooling PR 1441](https://github.com/apollographql/apollo-tooling/pull/1441) removes the `as!` when the property is a double-optional. Please head over there and let me know if you have any further questions. Thanks!

**Comment by designatednerd** [comment]
Fix to prevent double-optionals using `as!` shipped with `0.14.0`. If further problems arise after using that version, please open a new issue. Thank you!
